src/Shared/FileUtilities.cs

```diff
@@ -251,6 +253,75 @@ internal static String GetDirectoryNameOfFullPath(String fullPath)
             return null;
         }
 
+        /// <summary>
+        /// Gets the canonicalized full path of the provided path.
+        /// Path.GetFullPath The pre .Net 4.6.2 implementation of Path.GetFullPath is slow and creates strings in its work.
+        /// Therefore MSBuild has its own implementation on full framework.
+        /// Guidance for use: call this on all paths accepted through public entry
+        /// points that need normalization. After that point, only verify the path
+        /// is rooted, using ErrorUtilities.VerifyThrowPathRooted.
+        /// ASSUMES INPUT IS ALREADY UNESCAPED.
+        /// </summary>
+        internal static string NormalizePath(string path)
+        {
+            ErrorUtilities.VerifyThrowArgumentLength(path, "path");
+            string fullPath = GetFullPath(path);
+            return FixFilePath(fullPath);
+        }
+
+        internal static string NormalizePath(string directory, string file)
+        {
+            return NormalizePath(Path.Combine(directory, file));
+        }
+
+#if !CLR2COMPATIBILITY
+        internal static string NormalizePath(params string[] paths)
+        {
+            return NormalizePath(Path.Combine(paths));
+        }
+#endif
+
+        private unsafe static string GetFullPath(string path)
```

Comment:
Can you scope the `unsafe` to a block within the method just to reduce the surface area?

Reply:
I can extract that piece outside the method. Seems like if I block scope it then variable scoping ends up making the unsafe block still cover most of the method.